### PR TITLE
Speak Symbols When moving by word

### LINKED ISSUE #10855: Move the cursor one word to the right, symbol "|" not reported
This case occurs in NVDA_2019.3.1. While pressing ctrl + right arrow to move the cursor one word to the right, symbols are not reported. E.G. "|"

### Steps to reproduce:
1. Open notepad.
2. Paste text 
| 0 | 1 | 2 | 3 |
3. Press ctrl + right arrow several times to move the cursor one word to the right.
#### Actual behavior:
Symbols "|" are not reported.
#### Expected behavior:
NVDA reports "bar"

#### NVDA Installed/portable/running from source:
Installed. / portable

#### NVDA version:
2019.3.1

#### Windows version:
Windows version 10.0.18363

#### Does the issue still occur after restarting your PC?
Yes.

#### Have you tried any other versions of NVDA?
NVDA_2019.2.1 works as expected.

#### If addons are disabled, is your problem still occuring?
Yes.

#### Did you try to run the COM registry fixing tool in NVDA menu / tools?
Yes.

### LINKED ISSUE #11779: Inconsistencies in symbol pronunciation when reading by word and encountering a "word" which only contains multiple symbols.
### Steps to reproduce:

1. Open Word, or NotePad++ and type the line of text:
He said (quietly), "Hello, Jim".
Note the parenthesis either side of quietly, followed by a comma, and the quotes either side of "Hello Jim" followed by a full stop.
2. Press NVDA+P until the symbol level is set to "none".
3. Press HOME to move to the start of the line, then use NVDA+right arrow to move through the sentence by word.

### Actual behavior:

NVDA sends different information to the synthesizer, depending on the context of a symbol.  This can result in, for instance "left paren" being read, but not its right counterpart.  Or in the case of the quote, it is read once (where it is followed by a word), but not the second time (where it is followed by a full stop).
When moving by word, NVDA sends the following to the synthesizer (copied from the speech viewer:

He
said 
left paren
quietly
), 
quote
Hello
comma
Jim
".
blank

Because NVDA encounters the left paren and first quote as a single symbol on their own, the text "left paren" and the first "quote" are sent as words, and they are read.  Because ), and ". are sent as symbols (which are not set to be read at symbol level none) they are not read.

### Expected behavior:

NVDA should consistently read symbols (and either read them or not - but not read some and not others at the same symbol level).  Left and right paren are both symbol level "most" by default, as is quote.

#### NVDA version:

NVDA 2020.3

#### Windows version:

Windows 10 (64-bit) Version: 2004 Build 19041 

#### Name and version of other software in use when reproducing the issue:

Office 365 (64-bit) Version: 16.0.13231.20372 
NotePad++ v7.9

#### Other information about your system:

I was going to use Notepad, but it moves by word differently.  Where Notepad++ and Word both move to the start of the next word or group of symbols (a group may be one), Notepad simply moves to the first character after the next space, so doesn't read any characters in this case.  So, in Notepad, symbols are read only at the level they are set to read at - just as in Word or Notepad++ using Say all.  When moving by one character at a time, every character is read regardless of symbol level, as expeced.  The problem is when moving by word, and the whole "word" to be read consists of more than one symbol, they are treated by their symbol level.

I think the best solution is to ignore symbol level when moving by word.  Word is granular enough that it is worth knowing all the symbols you pass.  If reading by line or greater, then read according to symbol level.  The only case caught in the middle then would be a case where a line contains only symbols, which is a much rarer case than encountering a "word" of only symbols when moving by word.

#### Does the issue still occur after restarting your computer?
Yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
I don't think this behaviour has changed recently.  I tried NVDA 2019.2.1 portable and it behaves the same.

#### If addons are disabled, is your problem still occuring?
Yes
#### Did you try to run the COM registry fixing tool in NVDA menu / tools?
No

### PR DESCRIPTION
**Pull request by feerrenrut**
### Closed reason.
This requires an API breaking change, time ran out prior to the 2022.1 release while resolving usage and design concerns.
- There are many translated strings in NVDA (approx. 2300). It will take a while to consider them, and wrap with a type.
- The full set of types required is not known, what about strings with placeholders. The placeholder text may need to be considered for symbol substitution, but not the surrounding text. Consider the contrived example: `_("The column's name is %s")`. The apostrophe should not be announced as `tick`, but if the the column name should be processed eg for column name `don't`: `"The column's name is don tick t"`
- The commands inherit from `str` to ensure that they can be used in the same way is prior items in the speech sequence, however this also means it is easy for a logic error to drop the extra information provided by a class.

### Link to issue number:
https://github.com/nvaccess/nvda/pull/12695
Fixes: #10855
Fixes #11779

### Summary of the issue:
NV Access took open this abandoned PR:
Recent history: PR #12695 reverted PR #11856 fixing a regression described in issue #12653.

Navigation by word is inconsistent among different editors.
This might lead to different symbol announcements, or opening and closing symbols not announced consistently.

### Description of how this pull request fixes the issue:
- New config "symbolLevelWordAll" enabling symbol level ALL when navigating by word.
- When navigating by character, use symbol level all.
- When describing a symbol a new speech command is used to wrap the replacement.
  This propagates the initial intention and semantics of the speech sequence item.

Intentionally changes behavior for symbol text sent to synth when moving by word. This now matches the behavior for moving by line. The hyphen character in symbols like "t-shirt" and "right-pointing arrow" are sent to the synthesizer.

Note, commits have been carefully authored to show the progression of changes, however there are two commits that can be ignored, which only tidy related code:
- Clarify symbol level default
- add type information

The notepad tests have been confirmed to pass for all commits, looking at the changes to these tests will demonstrate the changes in behavior with that commit.

### Testing strategy:
- System tests
- Manual testing

### Known issues with pull request:
None

### Change log entries:
Already included in PR, see diff.

### Code Review Checklist:

- [x] Pull Request description is up to date.
- [x] Unit tests.
- [x] System (end to end) tests.
- [x] Manual testing.
- [x] User Documentation.
- [x] Change log entry.
- [x] Context sensitive help for GUI changes.
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers

### CODE REVIEW

**Comment by feerrenrut** [comment]
From the known issues with this PR:
>The final commit "Restore initial behavior for dash containing symbols" is potentially unnecessary.
However, a decision must be made about the preference for "t shirt" or "t-shirt" sent to the synth.
In the latter case a user's synth may speak this as "t dash shirt".

The changes in commit "Restore initial behavior for dash containing symbols":
https://github.com/nvaccess/nvda/pull/12710/commits/16da2f094dc7b2b6d1ed1d3fae082d7e5edb54ed#diff-a29456c11e438847c67bfe5c503ef7ca66bd0ffad420b23f7cf064505622cd45L1337

Restores the removed hyphen behavior.

We believe this code historically was required to provide the expected behavior for replacing whitespace characters (via the symbols pronunciation mechanism at level "character").

To resolve the "known issue" tests have been updated to include white space characters . The tests are added to the first commit of this PR to make it easier to assess the changes in behavior with each commit.

Reviewing the changes in `notepadTests.py` demonstrates how the behavior is changed.
I think we should drop the final commit, allowing the right pointing arrow and t shirt symbols to be spoken with hyphens in the text sent to the synthesizer.

**Comment by amirsol81** [comment]
Does this also apply to selecting text when SHIFT is added to navigation shortcuts?

**Comment by CyrilleB79** [comment]
I thought I had already commented in this PR but it seems not. So here I go.

I did not read it all carefully. However I can affirm that with French OneCore (Hortense), passing "T shirt" (without dash) to the synth instead of "T-shirt" (with dash) leads to an erroneous mispronunciation. Actually, in the latter case, the letter T is pronounced "tee" in English whereas in the first case, it is pronounced "teh" (in French) what is wrong since in the word "T-shirt" the "T" is pronounced  in English even in the French word.

**Comment by feerrenrut** [comment]
Thanks for the information @CyrilleB79. The systems tests for this PR show that before any changes, when moving by word in notepad, the synth is sent "t shirt" (no dash). Can you confirm that recent releases of NVDA using French OneCore this problem exists?

At this stage I expect we'll remove the code that results in the dashes being removed, but since it is a change in behavior I'm being cautious.

**Review by seanbudd — CHANGES_REQUESTED** [review]
Great work here, the general approach and testing looks good to me.

File: user_docs/en/changes.t2t
Comment (feerrenrut) [reply user_docs/en/changes.t2t:63]: Good catch.

File: tests/system/robot/notepadTests.robot
Comment (feerrenrut) [reply tests/system/robot/notepadTests.robot:22]: It was an artifact from debugging, but it can be helpful even when the tests pass and doesn't really have any drawbacks. I think it's best to leave it in.

File: tests/system/robot/notepadTests.py
Comment (feerrenrut) [reply tests/system/robot/notepadTests.py:19]: I don't think this is necessary for `BuiltIn`, we don't import it with an underscore in other files.

File: tests/system/robot/notepadTests.py
Comment (feerrenrut) [reply tests/system/robot/notepadTests.py:21]: I'm not sure it is necessary, but it makes it consistent. I'll update it.

File: source/speech/speech.py
Comment (feerrenrut) [reply source/speech/speech.py:370]: To be honest I'm not sure. I just broke it up into shorter lines and added return type information.
I don't think I could reword this quickly.

**Comment by feerrenrut** [comment]
I'll squash these latest commits into the branch once this is approved.

**Comment by lukaszgo1** [comment]
@feerrenrut  What I'm missing from the description of this PR is a reassurance for add-on authors that introduction of these new speech commands is not going to cause breakage for them. Could you please describe any testing that you've done with add-ons (particularly with ones which were causing troubles with introduction of the new speech commands such as NVDA Remote)?

**Review by seanbudd — APPROVED** [review]
Thanks @feerrenrut, LGTM pending confirmation that add-ons are not broken

**Comment by feerrenrut** [comment]
@lukaszgo1 No, this hasn't been looked at, good catch.

It certainly is a tricky one. Our intention certainly is not to break add-ons, but the cost could be a more complicated codebase (full of workarounds), or being unable to take up new features / fix issues (such as this PR) until a point one release.  Since these ARE speech commands, they should be handled (and likely ignored) by addons. Granted, for addons (such as NVDA remote) that need to process speech, loosing this speech would be a problem.

I'll look into whether this affects NVDA remote, and also consider how we can alter the approach.

**Comment by feerrenrut** [comment]
After some investigation into different approaches, it will be very difficult to introduce this feature without introducing a lot of complexity, or potentially degrading addons.

Instead, we'll postpone this feature until the 2022.1 release.

**Comment by CyrilleB79** [comment]
> Thanks for the information @CyrilleB79. The systems tests for this PR show that before any changes, when moving by word in notepad, the synth is sent "t shirt" (no dash). Can you confirm that recent releases of NVDA using French OneCore this problem exists?

@feerrenrut, no, surprisingly, this problem does not exist. I have been using nvda_snapshot_alpha-23630,0f481d8a.exe and I have tested with French OneCore (Hortense) in Notepad. The results are:
* the '👕' symbol is read correctly either by line, by character or by word
* a line containing "T-shirt" (with dash) is read correctly
* a line containing "T shirt" (without dash) is NOT read correctly

I have activated the log of SpeechManager debug messages and saw something very strange:
* With a French voice, the symbol is spelt
  * "T-shirt" (with dash) at IO log level
  * "T-shirt  " (with dash and with two trailing spaces) at SpeechManager debug log level.
* With an English voice, the symbol is spelt:
  * "t-shirt" (with dash) at IO log level
  * "t shirt  " (without dash and with two trailing spaces) at SpeechManager debug log level.

I have no explanation why the dash is removed in English but not in French. But this makes the T-shirt read correctly by chance...

**Review by Qchristensen — APPROVED** [review]
I don't have an answer to the French pronunciation of T-Shirt either - does it vary when testing with different synthesizers?  (which would likely indicate a pronunciation issue with one synth).  Just trying to think of other similar things... T-Rex, the dinosaur, or T-Ball, the kids sport.

Otherwise, updates to changes and user guide look fine.

**Comment by lukaszgo1** [comment]
> 
> 
> It would be great to introduce French variations of the system tests. Hopefully NVDA isn't affected too heavily by the Windows language. Though system tests are a good way to find that.


Is there any interest in general in making system  tests working on non English versions of Windows - for now they don't work at all.

**Comment by feerrenrut** [comment]
Rather than specifically a French version of windows, I would start with NVDA set to French language for certain system tests that exercise aspects that may be broken more easily (EG symbol processing, decimal, perhaps dates)?. Ideally this should work on any machine, we should invert the dependency on Windows API calls so that it can be controlled during a system test.

**Comment by feerrenrut** [comment]
Hi @CyrilleB79 I've got time to look at this again. Thanks for providing test cases.

First a summary of my findings, I looked at test 1 and test 2 [from your comment on this PR](https://github.com/nvaccess/nvda/pull/12710#issuecomment-949963761)

### Test 1:
Yes, this looks like one of the situations that the [new speech commands](https://github.com/nvaccess/nvda/pull/12710/files#diff-9e4e5cf381b5f223fbfa727130be1471a16eef6f7ce80a68f3337645d2c5b47fR52) ([usage](https://github.com/nvaccess/nvda/pull/12710/files#diff-a29456c11e438847c67bfe5c503ef7ca66bd0ffad420b23f7cf064505622cd45R285)) were introduced to allow us to differentiate between "content" that should be subject to speech processing rules, and "speech UI" that should be fixed.
We'll have to decide if it is appropriate to go through the code base to specifically categorize all translated string to prevent further errors like this.

### Test 2:
The actual behavior matches my expectations, please let me know if you disagree.

>Note also that this issue would disappear if you follow the proposal made in https://github.com/nvaccess/nvda/pull/11856#issuecomment-739914763. Would you re-consider it?

I think you are referring to the option "Speak all symbols when reviewing by word".
This PR includes that option

### Test 2 details
I have shortened the example to highlight the behavior:
Example:
He didn't.

Symbol level: None, 
Speak all symbols when reviewing by word: yes
By line: he didnt
By word: he didn tick t dot
By char: h e space d i d n tick t dot

Symbol level: None, 
Speak all symbols when reviewing by word: no
By line: he didnt
By word: he didnt
By char: h e space d i d n tick t dot

Symbol level: All,
Speak all symbols when reviewing by word: no
By line: he didn tick t dot
By word: he didn tick t dot
By char: h e space d i d n tick t dot

File: source/speech/speech.py
Comment (lukaszgo1) [inline source/speech/speech.py:1162]: Please disregard my last comment - wrapping the text in the `Symbol` command in `editableText` would not improve the situation at all. In this line you're overwriting symbol level for the entire sequence passed to `speech.speak` which I believe is causing the unnecessary punctuation to be announced. This code should be changed so that  the punctuation level is changed only for the duration of announcing the word / character and not for the entire sequence This would also eliminate the need for the new `UserInterface` command decreasing the complexity so this is a good thing.

Comment (feerrenrut) [reply source/speech/speech.py:1162]: > This code should be changed so that the punctuation level is changed only for the duration of announcing the word / character and not for the entire sequence.

How would you do that?

**Comment by feerrenrut** [comment]
>First of all, I can tell that with the new snaphot of today, I confirm that spelling erorrs in French do not cause any issue anymore.

Great. Though, at this stage it is expected there will be many others, I should have made that clear in my last comment.

> We'll have to decide if it is appropriate to go through the code base to specifically categorize all translated string to prevent further errors like this.

>all translatable strings of the code base should be processed the same way as "spelling error",

Not quite, translated strings are used in the GUI or in Braille, and they are also used for values consumed by both (eg names of roles / states). If used everywhere we could wrap the string as part as the translation function.

There are 2333 translatable strings in the code base that I found with: 
`git grep -o \_\([\"\'] | wc -l`

Some directories could be initially excluded. EG gui with 557 strings

>This seems unmanageable - there would always be one more message we missed for some language.

Rather than do this based on a test then change approach. Instead, we'd go through all strings and ensure they are commands when used for speech. The proposal is here: https://github.com/nvaccess/nvda/issues/12779

I'll look into the alternatives. However, there isn't much time left to include this in 2022.1.

**Comment by lukaszgo1** [comment]
> Not quite, translated strings are used in the GUI or in Braille, and they are also used for values consumed by both (eg names of roles / states). If used everywhere we could wrap the string as part as the translation function.

This would not solve the issues @CyrilleB79  raised above i.e:
- Name of the style containing punctuation when navigating by word
- content of the table column header containing punctuation.

> I'll look into the alternatives. However, there isn't much time left to include this in 2022.1.

I believe this cannot be merged as is - there is too much broken cases right now.

File: source/speech/speech.py
Comment (lukaszgo1) [reply source/speech/speech.py:1162]: The responsibility of deciding what symbol level to use should be moved from `speakTextInfo` (this function has not enough information to make a right decision as it receives only a sequence of strings) to the `getTextInfoSpeech`. `getTextInfoSpeech` should retrieve list of all fields and strings  in the textInfo using `getTextWithFields` as it does now, and then should enclose all strings in a speech command indicating to `speech.speak` with which punctuation level the given string should be spoken.

**Comment by CyrilleB79** [comment]
Here is my feedback on Test 2

You wrote:
> ### Test 2:
> 
> The actual behavior matches my expectations, please let me know if you disagree.
> 
> > Note also that this issue would disappear if you follow the proposal made in [#11856 (comment)](https://github.com/nvaccess/nvda/pull/11856#issuecomment-739914763). Would you re-consider it?
> 
> I think you are referring to the option "Speak all symbols when reviewing by word". This PR includes that option
> ### Test 2 details
> 
> I have shortened the example to highlight the behavior: Example: He didn't.
> 
> Symbol level: None, Speak all symbols when reviewing by word: yes By line: he didnt By word: he didn tick t dot By char: h e space d i d n tick t dot
> 
> Symbol level: None, Speak all symbols when reviewing by word: no By line: he didnt By word: he didnt By char: h e space d i d n tick t dot
> 
> Symbol level: All, Speak all symbols when reviewing by word: no By line: he didn tick t dot By word: he didn tick t dot By char: h e space d i d n tick t dot

I agree that this behavior is the one observed with the snapshot of this PR. Let's call it NVAccess proposal.

In the various comments and issues/PRs on this topic, I have expressed my opinion on this behavior. Here is a summary:

1. Initially, I would have preferred to have only symbols at the beginning and end of word controlled by the new option, in-word symbols being controlled by the existing general symbol level. With your example, that means:
   Symbol level: None,
   Speak all symbols when reviewing by word: yes
   By line: he didnt
   By word: he didnt dot
   By char: h e space d i d n tick t dot

2. NVAccess have indicated (e.g. in https://github.com/nvaccess/nvda/pull/11856#issuecomment-739700725) that implementing this may be confusing.

3. Thus I agree that NVAccess' proposal is more clear than mine and support merging it even if it does not solve all the issues.

4. I have still however concerns regarding in-word symbols being pronounced too much, for my own use as well as for other languages if some of them use more frequently the tick in the middle of words. Thus I think that it should be tested well by various foreignn communities during alpha and beta stage.

Given the point 4., I think there is a risk to have this PR reverted again. Maybe it would be wise if possible to split this PR in two separate PRs: one addressing the API-breaking changes, e.g. new speech commands, and the other one dedicated to symbol pronunciation when reading by word more specifically. Would it be doable?

**Comment by CyrilleB79** [comment]
An additional point to be considered: are the symbols reported when selecting by word with ctrl+shift+left/rightArrow? What is expected?

In the last test I have done, I think that it was not reported. But the test was not done on the last commit; to be checked...

**Comment by feerrenrut** [comment]
Hi @CyrilleB79, I think this PR is getting a bit off track. Some of the issues raised are pre-existing issues (see #13305), I agree they are important, but I want to bring the focus back to the core of this issue:

To summarize where we are at:
- The core is to fix #10855
 (Pipe symbol not reported when moving by word) and #11779 (Inconsistencies in symbol pronunciation... )
- Additionally, based on the concern that not everyone wants "symbol level all" for move by word, an option was added to fallback to the prior intended behavior of observing the symbol level setting.
- We have also fixed some of the concerns demonstrated in #13305. Specifically, double processing of symbols (EG t-shirt and right-pointing-arrow), and the core mechanism to resolve symbol substitution within the speech UI messages.

From here, I'll wrap (certain obvious) speech messages with a UI speech command, and try to get this merged.
Rather than aim for a complete fix for this pre-existing issue, I'll improve the obvious case, leave a mechanism for improving the rest. I don't want this PR to get held up further fixing pre-existing issues that weren't in the initial scope.

**Comment by CyrilleB79** [comment]
Hi @feerrenrut 

Thanks for this summary.

> * The core is to fix [Move the cursor one word to the right, symbol "|" not reported #10855](https://github.com/nvaccess/nvda/issues/10855)
>   (Pipe symbol not reported when moving by word) and [Inconsistencies in symbol pronunciation when reading by word and encountering a "word" which only contains multiple symbols. #11779](https://github.com/nvaccess/nvda/issues/11779) (Inconsistencies in symbol pronunciation... )

Actually #10855 is already fixed in 2020.3 as explained in https://github.com/nvaccess/nvda/issues/10855#issuecomment-1028805088. So this PR just fixes #11779.

> * Additionally, based on the concern that not everyone wants "symbol level all" for move by word, an option was added to fallback to the prior intended behavior of observing the symbol level setting.

Yes, this should be enough to mitigate the risk of reverting this PR (on the contrary to what I have written before).

> * We have also fixed some of the concerns demonstrated in [add tests for symbol levels #13305](https://github.com/nvaccess/nvda/pull/13305). Specifically, double processing of symbols (EG t-shirt and right-pointing-arrow), and the core mechanism to resolve symbol substitution within the speech UI messages.
> 
> From here, I'll wrap (certain obvious) speech messages with a UI speech command, and try to get this merged. Rather than aim for a complete fix for this pre-existing issue, I'll improve the obvious case, leave a mechanism for improving the rest. I don't want this PR to get held up further fixing pre-existing issues that weren't in the initial scope.

Yes this makes sense. This PR should be merged as soon as possible in order to identify if some more messages should be wrapped in UI speech command during alpha or beta testing phase.

Just one last point: you haven't answered to https://github.com/nvaccess/nvda/pull/12710#issuecomment-1021682480, i.e. symbols are not reported when selecting by word.
It should be clarified:
* if you plan to fix it in this PR
* if you consider it an issue but do not plan to fix it in this PR
* if you do not consider it an issue

**Comment by lukaszgo1** [comment]
I still stand by my opinion that this should not be merged as is.
To avoid getting out of scope I'm not going to discuss what symbol level should be used when selecting and I'll focus on the regressions this PR introduces mainly based on [this comment](https://github.com/nvaccess/nvda/pull/12710#issuecomment-1021079467)

> 
>     * Additionally, based on the concern that not everyone wants "symbol level all" for move by word, an option was added to fallback to the prior intended behavior of observing the symbol level setting.

While this works for word navigation this PR also changes the behavior when navigating by character. The entire sequence is spoken at a level all which causes punctuation in names of the table header and in style names to be wrongly reported and this cannot be controlled by the user.

> and the core mechanism to resolve symbol substitution within the speech UI messages.

As demonstrated by the comment to which I've linked above this mechanism is not good enough. While you would be able to fix style names not being reported with the level all by enclosing them in the right speech command the same cannot be done for names of table headers.

What worries me is that I see no way to improve the situation without introducing additional backwards incompatible changes meaning that any fixes would have to wait until 2023.1.
Since we have identified certain regressions which would be introduced when this PR is going to be merged, the proposed mechanism for excluding strings from being spoken at level all is clearly not flexible enough, we do not have a clear path forward for resolving issues with the entire sequences being spoken at the wrong level and there is no high demand for #11779 to be implemented soon I believe a right thing to do is not to include this half baked in 2022.1 and rework the implementation for 2023.1

**Comment by CyrilleB79** [comment]
> To avoid getting out of scope I'm not going to discuss what symbol level should be used when selecting 

Sorry, but IMO it is not out of scope of this PR.
Indeed, currently, when moving by character, symbols are reported whatever the symbol level is. And selecting by character, the symbol is also reported whatever the symbol level is.
With this PR, moving by word, symbols will be reported whatever the symbol level is. It would seem coherent to me that with this PR, selecting by word, symbol be reported whatever the symbol level is.

> While this works for word navigation this PR also changes the behavior when navigating by character. The entire sequence is spoken at a level all which causes punctuation in names of the table header and in style names to be wrongly reported and this cannot be controlled by the user.

Sorry, I had missed this point. I completely agree with you. The use case moving by character in a table with headers containing symbols is a real-life use case and merging this PR as is would introduce a visible regression.

**Comment by lukaszgo1** [comment]
> Sorry, but IMO it is not out of scope of this PR. Indeed, currently, when moving by character, symbols are reported whatever the symbol level is. And selecting by character, the symbol is also reported whatever the symbol level is. With this PR, moving by word, symbols will be reported whatever the symbol level is. It would seem coherent to me that with this PR, selecting by word, symbol be reported whatever the symbol level is.

I agree that currently this  is inconsistent, but there is no visible change in the behavior when selecting text, this can be worked on in a follow up PR and it is not necessary to fix #11779.

**Comment by feerrenrut** [comment]
I have tested table headers with a [codepen sample](https://codepen.io/reefturner/pen/KKyggoO) and didn't find any situation that had been made worse. Ideally, add an automatic test case to demonstrate the "before" behavior on `master`.

We have exhausted the amount of time we can dedicate to this now. I'll spend a little time on on Monday to write up where this stands, document the approach, and clean up the open issues to clarify goals. I'll close this PR, anyone wishing to pick this up should open a new one.

**Comment by lukaszgo1** [comment]
> I have tested table headers with a [codepen sample](https://codepen.io/reefturner/pen/KKyggoO) and didn't find any situation that had been made worse.

You can reproduce the regression as follows:
- Position yourself in the cell with the text "Blue" inside
- Move with the right arrow to the cell containing "True" and notice that the column header "don't" is announced as "don tick t"

> 
> We have exhausted the amount of time we can dedicate to this now. I'll spend a little time on on Monday to write up where this stands, document the approach, and clean up the open issues to clarify goals. I'll close this PR, anyone wishing to pick this up should open a new one.



While dropping this work after so much time has been spend on it must have been difficult I really applaud you for doing a right thing here! The fact that we can merge code when it is working as best as possible rather that on some arbitrary deadline is one of the main reasons why I find working on Open Source stuff so rewarding personally.

**Comment by feerrenrut** [comment]
Let's be clear about the behavior with the table:
In 2021.3.1:
- with symbol level all, `don tick t` reported
- with symbol level none, `don't` reported
With this PR:
- with symbol level None and "symbol level all for move by word" disabled, `don tick t` reported.

This case is questionable, some thoughts on the matter:
- The intention with this PR is to speak at symbol level all consistently, it's doing that.
- However, in this case it is not primary content being navigated over, it is secondary (IE context for the cell).
- It wouldn't be particularly difficult to work around in a similar way to the speech UI, a new command for "secondary" or "context" content. 

Note, with the approach demonstrated here, and with #12779 the goal would be to have no un-annotated strings in speech sequences anymore. This would require introducing a "Content" type (in addition to UserInterface, and Symbol). There are few different approaches that we can consider:
- We could attach metadata to the these types, like `symbolLevel`. This results in the creation code (rather than the type) having control. One `Content` instance may have a different result from another `Content` instance.
- We could add more types for the different behaviors eg (`Context` could be a type of `Content` which doesn't expand symbols). The calling code still has control (based on which type is created), but once created it has consistent behavior.

**Comment by feerrenrut** [comment]
If this gets picked up again, I'd recommend the first step be to rebase onto https://github.com/nvaccess/nvda/pull/13305 and confirm the output with those tests.

**Comment by CyrilleB79** [comment]
Hi @feerrenrut

Since this one is an API-breaking change, do NVAccess devs plan themselves to implement it for 2023.1 release? Or do you need some help for this?

This was initially planned for 2022.1 but was associated with #11779's fix and was reverted due to issues linked with #11779's fix.

Probably this issue should be fixed for 2023.1 with API-breaking changes. Then #11779 could be resolved in 2023.1 or a later release.

Could you give an update of NVAccess' opinion on the resolution of this issue? Thanks.

**Comment by seanbudd** [comment]
@CyrilleB79, we do have plans to pick this up again for 2023.1. However, it is a large and risky piece of work, that might not make it in again. Having assistance reviving this PR would be helpful. At the very least it would get momentum on this sooner and increase chances of getting this change in 2023.1. It would be great to address any remaining issues in the PR, and get testing as soon as possible. This will require a large breadth of testing. It would be good to cover smoke tests, system and unit tests for changed code.
I suggest continuing work by opening PRs against `speakSymsMoveByWord`, which will become a feature branch via #14287

**Pull request closed without merging**
